Sources/Db/DatabaseApi.php

```diff
@@ -0,0 +1,530 @@
+	 * always be a reference to the first instance's local $connection object.
+	 * It will not be overwritten by later instances.
+	 */
+	public static $db_connection;
+
+	/**
+	 * @var bool
+	 *
+	 * Force unbuffered output. Only applicable to MySQL.
+	 */
+	public static bool $unbuffered = false;
+
+	/**********************
+	 * Protected properties
+	 **********************/
+
+	/**
+	 * @var array
+	 *
+	 * SMF tables that can't be auto-removed - in case a mod writer cocks it up!
+	 */
+	protected array $reservedTables = array(
+		'admin_info_files',
+		'approval_queue',
+		'attachments',
+		'background_tasks',
+		'ban_groups',
+		'ban_items',
+		'board_permissions',
+		'board_permissions_view',
+		'boards',
+		'calendar',
+		'calendar_holidays',
+		'categories',
+		'custom_fields',
+		'group_moderators',
+		'log_actions',
+		'log_activity',
+		'log_banned',
+		'log_boards',
+		'log_comments',
+		'log_digest',
+		'log_errors',
+		'log_floodcontrol',
+		'log_group_requests',
+		'log_mark_read',
+		'log_member_notices',
+		'log_notify',
+		'log_online',
+		'log_packages',
+		'log_polls',
+		'log_reported',
+		'log_reported_comments',
+		'log_scheduled_tasks',
+		'log_search_messages',
+		'log_search_results',
+		'log_search_subjects',
+		'log_search_topics',
+		'log_spider_hits',
+		'log_spider_stats',
+		'log_subscribed',
+		'log_topics',
+		'mail_queue',
+		'member_logins',
+		'membergroups',
+		'members',
+		'mentions',
+		'message_icons',
+		'messages',
+		'moderator_groups',
+		'moderators',
+		'package_servers',
+		'permission_profiles',
+		'permissions',
+		'personal_messages',
+		'pm_labeled_messages',
+		'pm_labels',
+		'pm_recipients',
+		'pm_rules',
+		'poll_choices',
+		'polls',
+		'qanda',
+		'scheduled_tasks',
+		'sessions',
+		'settings',
+		'smiley_files',
+		'smileys',
+		'spiders',
+		'subscriptions',
+		'themes',
+		'topics',
+		'user_alerts',
+		'user_alerts_prefs',
+		'user_drafts',
+		'user_likes',
+	);
+
+	/***********************
+	 * Public static methods
+	 ***********************/
+
+	/**
+	 * Static method used to instantiate the child class specified by Config::$db_type.
+	 *
+	 * If $options is empty, correct settings will be determined automatically.
+	 *
+	 * @param array $options An array of database options.
+	 * @return object An instance of a child class of this class.
+	 */
+	final public static function load(array $options = array())
+	{
+		if (isset(self::$db))
+			return self::$db;
+
+		// Figure out what type of database we are using.
+		switch (!empty(Config::$db_type) ? strtolower(Config::$db_type) : 'mysql')
+		{
+			// PostgreSQL is known by many names.
+			case 'postgresql':
+			case 'postgres':
+			case 'postgre':
+			case 'pg':
+				$class = POSTGRE_TITLE;
+				break;
+
+			// MySQL and its forks.
+			case 'mysql':
+			case 'mariadb':
+			case 'percona':
+				$class = MYSQL_TITLE;
+				break;
+
+			// Something else?
+			default:
+				$class = ucwords(Config::$db_type);
```

Comment:
Be careful with this.  If setlocale has been called, this may not work like you expect because not all languages capitalize like english and it would be ignored.  If we don't have setlocale called it should be safe, but SMF does/did set this.

Reply:
It should be fine. `SMF\Lang::load()` is the only place where we call `setlocale()`, and `SMF\Lang::load()` is never called until long after we've loaded the database.